### PR TITLE
feat: infer typing from celebrate args

### PR DESCRIPTION
**Pull request by alburkerk**
Related issue: 
https://github.com/arb/celebrate/issues/212

### CODE REVIEW

**Comment by arb** [comment]
I don't know TS that well... what does this change get us specifically. Like not from a type soundness standpoint but what functional change does this get us?

**Comment by alburkerk** [comment]
Currently there is:

```typescript
interface Paging {
    limit: number;
    page: number;
}

let paging = Joi.object<Paging>({
    limit: Joi.number();
    page: Joi.number();
});

router.get('/', celebrate({ query: paging }), async (req, res, next) => {
    const query = req.query; // type QueryString.ParsedQS (default from express)
})
````

In order to have correct typing, one must write:

```typescript
router.get('/', celebrate<ParamsDictionary, any, any, Paging>({ query: paging }), async (req, res, next) => {
    const query = req.query; // type Paging
})
```

This a boring notation, even more because we need put all 4 generic values just to type the generic query type.

With my changes, the first notation works and we have:

```typescript
router.get('/', celebrate({ query: paging }), async (req, res, next) => {
    const query = req.query; // type Paging
})
```

Because type is inferred from arguments

File: lib/index.d.ts [hunk 1]
```diff
@@ -4,6 +4,7 @@ import {
     Root as joi,
     ValidationOptions,
     ValidationError,
+    default as JoiDefault
```
Comment (arb) [inline lib/index.d.ts:7]: Is there any way we can reduce the number of imports from `joi` here? What is the difference between `joi` and `JoiDefault` here?

Comment (alburkerk) [reply lib/index.d.ts:7]: Could indeed import Schema instead of JoiDefault, will modify the PR soon

**Comment by arb** [comment]
@alburkerk are you still interested in landing this PR?

**Pull request closed without merging**